luigi/s3.py

```diff
@@ -472,12 +482,20 @@ def listdir(self, path, start_time=None, end_time=None):
                     (not start_time and end_time and last_modified_date < end_time) or  # end defined, prior to end
                     (start_time and end_time and start_time < last_modified_date < end_time)  # both defined, between
                ):
-                yield self._add_path_delimiter(path) + item.key[key_path_len:]
+                if with_key:
+                    yield self._add_path_delimiter(path) + item.key[key_path_len:], item
+                else:
+                    yield self._add_path_delimiter(path) + item.key[key_path_len:]
 
     def list(self, path, start_time=None, end_time=None):  # backwards compat
         key_path_len = len(self._add_path_delimiter(path))
-        for item in self.listdir(path, start_time=start_time, end_time=end_time):
-            yield item[key_path_len:]
+        for path in self.listdir(path, start_time=start_time, end_time=end_time):
+            yield path[key_path_len:]
+
+    def list_with_key(self, path, start_time=None, end_time=None):
+        key_path_len = len(self._add_path_delimiter(path))
+        for path, item in self.listdir(path, start_time=start_time, end_time=end_time, with_key=True):
+            yield path[key_path_len:], item
```

Comment:
My spidey senses tell me this could be condensed to a single function with an extra argument, but I'm not too concerned by it